t/semaphore.t

```diff
@@ -0,0 +1,1068 @@
+post
+true
+post
+true
+wait
+true
+1
+1
+wait
+true
+2
+2
+--- no_error_log
+[error]
+
+
+
+=== TEST 21: benchmark
+--- http_config eval: $::HttpConfig
+--- config
+    location = /test {
+        access_log off;
+        content_by_lua '
+            local semaphore = require "ngx.semaphore"
+            local g = getmetatable(_G).__index
+            if g.id == nil then
+                g.id = 0
+                g.map = {}
+            end
+            local id = g.id
+            g.id = g.id + 1
+            local sem = semaphore.new(0)
+            g.map[id] = sem
+            local res1, res2 = ngx.location.capture_multi{
+                { "/sem_wait",{ method = ngx.HTTP_POST, body = tostring(id) }},
+                { "/sem_post",{ method = ngx.HTTP_POST,body = tostring(id)}},
+            }
+            g.map[id] = nil
+            ngx.say("ok")
+        ';
+    }
+
+    location /sem_wait {
+        content_by_lua '
+            local semaphore = require "ngx.semaphore"
+            local g = getmetatable(_G).__index
+            ngx.req.read_body()
+            local body = ngx.req.get_body_data()
+            local sem = g.map[tonumber(body)]
+            local ok, err = sem:wait(10)
+            if ok then
+                --ngx.print("wait")
+                ngx.exit(200)
+            else
+                ngx.log(ngx.ERR, err)
+                ngx.exit(500)
+            end
+        ';
+    }
+
+    location /sem_post {
+        content_by_lua '
+            local semaphore = require "ngx.semaphore"
+            local g = getmetatable(_G).__index
+            ngx.req.read_body()
+            local body = ngx.req.get_body_data()
+            local sem = g.map[tonumber(body)]
+            local ok, err = sem:post()
+            if ok then
+                --ngx.print("post")
+                ngx.exit(200)
+            else
+                ngx.log(ngx.ERR, err)
+                ngx.exit(500)
+            end
+
+    ';
+    }
+
+--- request
+GET /test
+--- response_body
+ok
+--- no_error_log
+[error]
+
+
+
+=== TEST 22: semaphore wait time is zero
+--- http_config eval: $::HttpConfig
+--- config
+    location /test {
+       content_by_lua '
+            local semaphore = require "ngx.semaphore"
+            local sem = semaphore.new(0)
+            local ok, err = sem:wait(0)
+            if not ok then
+                ngx.say(err)
+            end
+       ';
+}
+--- no_check_leak
+--- request
+GET /test
+--- response_body
+busy
+--- no_error_log
+[error]
+
+
+
+=== TEST 23: basic semaphore wait time default is zero
+--- http_config eval: $::HttpConfig
+--- config
+    location /test {
+       content_by_lua '
+            local semaphore = require "ngx.semaphore"
+            local sem = semaphore.new(0)
+            local ok, err = sem:wait()
+            if not ok then
+                ngx.say(err)
+            end
+       ';
+}
+--- no_check_leak
+--- request
+GET /test
+--- response_body
+busy
+--- no_error_log
+[error]
```

Comment:
I runed gco after you mention this. The rate of coverage in ngx_http_lua_semaphore.c  is 91%. Is it OK? I don't know much about covrage testing.